### ISSUE TITLE
Buffer's text is gone after spawning a new window in swaywm

### ISSUE THREAD

**Issue by kohane27** [neovim/neovim#24272]
### Problem

Hello there. Hope you're doing well. Thank you for checking out this bug issue of mine. I appreciate it.

A picture is worth a thousand words so please see the following video:

https://github.com/neovim/neovim/assets/57322459/a9a1946c-31e9-4314-b9f4-de298128dc3a

1. [kitty](https://sw.kovidgoyal.net/kitty/) is opened on workspace 1
2. I spawned a new terminal next to it
3. All the text in the nvim window is gone. But when I scroll, it re-appears
4. The `Enter` key is not responding, i.e., pressing `Enter` has no effect
5. I had to kill the terminal

Another video showing the key `Enter` has no response. The terminal is `foot`, showing it's not the terminal issue:

https://github.com/neovim/neovim/assets/57322459/53a640cb-e99d-4e39-92bc-0d7402518001

## What I Have Tried

1. Open nvim in a terminal without tmux, but the same issue persists
2. Likely similar or same issue mentioned in https://github.com/neovim/neovim/issues/24239 because the command `:resize` fixed it (the Enter key doesn't work but `C-j` (acts as Enter in the terminal) works)
3. My main terminal is [kitty](https://sw.kovidgoyal.net/kitty/). I tried other GPU-accelerated terminals like [WezTerm](https://wezfurlong.org/wezterm/) and [Alacritty](https://alacritty.org/). They all have the same issue. I thought it's the GPU aspect of it, so I also tried non-GPU-accelerated terminals like [foot](https://codeberg.org/dnkl/foot) and UXTerm. They also have the same issue.

```
:version
NVIM v0.9.1
Build type: Release
LuaJIT 2.1.0-beta3

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```

System info:

```
Kernel: Linux 6.1.37-1-lts
WM: sway
CPU: AMD Ryzen 7 PRO 5850U with Radeon Graphics (16) @ 1.900GHz
GPU: AMD ATI Radeon Vega Series / Radeon Vega Mobile Series
```

Any input is much appreciated. Thank you.

### Steps to reproduce

please see above video description

### Expected behavior

buffer's text is not gone and enter key works

### Neovim version (nvim -v)

NVIM v0.9.1

### Vim (not Nvim) behaves the same?

no, VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Jun 14 2023 00:27:53)

### Operating system/version

Linux 6.1.37-1-lts

### Terminal name/version

kitty 0.28.1

### $TERM environment variable

tmux-256color

### Installation

aur

**Comment by justinmk** [comment]
Can you try the development version? 

- The [Releases](https://github.com/neovim/neovim/releases/tag/nightly) page has pre-built archives for Linux/Windows/macOS. 
- Check `:version` to confirm that you are using the latest version.

**Comment by kohane27** [comment]
Hello @justinmk  , thank you for such a quick reply! I appreciate it.

I installed through the aur `neovim-git`:

```bash
:version
NVIM v0.10.0-dev-638+g4e34ca8ae7
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
```

And the issue is gone! Thank you so much!

If you have time I'd appreciate it if you could enlighten me about its cause.

I guess now I have to wait for this bugfix to be back ported to the stable version. Thank you again!

**Comment by kohane27** [comment]
Hello @justinmk ! Thank you for such a quick reply! I appreciate it.

I installed through the aur `neovim-git`:

```bash
:version
NVIM v0.10.0-dev-638+g4e34ca8ae7
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
```

Please see the video below:

https://github.com/neovim/neovim/assets/57322459/85eee305-028f-4073-b604-3317df083646

The issue doesn't immediately appear for the first spawn but it shows up on the second spawn of new terminal.

I'll provide a minimal config to see if the same issue exists.

Thank you again!

**Comment by kohane27** [comment]
I tried the minimal config [fzf-lua#quickstart](https://github.com/ibhagwan/fzf-lua#quickstart) and the issue is not showing up:

https://github.com/neovim/neovim/assets/57322459/71d0e702-aeaa-4cef-ac80-69dd7ab54e83

I guess it's one of my plugins causing this issue. I'll investigate and provide further comment. Thank you!

**Comment by kohane27** [comment]
I updated today with the version being `NVIM v0.10.0-dev-641+g811140e276`, and I'm very glad to report that this issue is gone!
I guess now I have to wait for this bugfix to be back ported to the stable version. Thank you again!

**Comment by justinmk** [comment]
?

**Comment by kohane27** [comment]
@justinmk Please see the video:

https://github.com/neovim/neovim/assets/57322459/4518f652-48d8-4e62-be19-67ec91cc7bc2

Apologies for my mistake: I didn't test it thoroughly enough.

After switching the nightly (currently on `NVIM v0.10.0-dev-643+g804108af6c`), this issue doesn't occur every time. I'll try to find a way to reliably reproduce this issue. Maybe it's my config issue, maybe it only happens on the second monitor. I am not sure...

I'll not close the issue again until I'm sure it's fixed. My bad.

**Comment by kohane27** [comment]
tl;dr: it's my `zsh` function causing the issue, not neovim.

It took me days to figure it out. It's not `linux`, `sway`, `tmux`, `ranger`, `kitty`, `wezterm` issue. The culprit is the following `zsh` function:

```bash
# very tenacious bug: causing terminal redraw issue
ranger_cd() {
    echo
    temp_file="$(mktemp -t "ranger_cd.XXXXXXXXXX")"
    ranger --choosedir="$temp_file" -- "${@:-$PWD}" <"$TTY"
    if chosen_dir="$(cat -- "$temp_file")" && [ -n "$chosen_dir" ] && [ "$chosen_dir" != "$PWD" ]; then
        cd -- "$chosen_dir" || exit
    fi
    zle reset-prompt
}

zle -N ranger_cd
bindkey '^F' ranger_cd
```

Whenever I use the above function to launch `ranger` to open `neovim`, the issue surfaces.

I've been using the above function for a long time, but suddenly it's causing this bug. I have no idea why. If, of the rare chance that someone is having the same problem as me, I'm now using the following substitute:

```bash
function r() {
  tempfile="$(mktemp -t tmp.XXXXXX)"

  ranger --choosedir="$tempfile" "${@:-$(pwd)}"

  test -f "$tempfile" && if [ "$(cat -- "$tempfile")" != "$(echo -n "$(pwd)")" ]; then
      cd -- "$(cat "$tempfile")" || return
      zle reset-prompt
    fi
}

bindkey -s "^F" "r\n"
```

Apologies for the trouble and noise. It's not neovim issue. I switched back to the stable release and the issue never re-surfaced after using the above function. Thanks guys!
